specs/data_structures.md

```diff
@@ -224,49 +225,91 @@ Merkle trees are used to authenticate various pieces of data across the LazyLedg
+Binary Merkle trees are constructed in the usual fashion, with leaves being hashed once to get leaf node values and internal node values being the hash of the concatenation of their children.
 
-For leaf node of message `m`:
+For leaf node of leaf message `m`, its value `v` is:
 ```C++
-v = h(0x00, serialize(m))
+v = h(serialize(m))
 ```
 
-An exceptions is made, in the case of empty leaf nodes: the value of an empty leaf node is 32-byte zero, i.e. `0x0000000000000000000000000000000000000000000000000000000000000000`. This is used rather than duplicating the last node if there are an odd number of nodes in order to avoid [CVE-2012-2459](https://nvd.nist.gov/vuln/detail/CVE-2012-2459). Implicitly, trees are padded with empty nodes up to the next larger power of 2.
+An exception is made, in the case of empty leaves: the value of a leaf node with an empty leaf is 32-byte zero, i.e. `0x0000000000000000000000000000000000000000000000000000000000000000`. This is used rather than duplicating the last node if there are an odd number of nodes in order to avoid [CVE-2012-2459](https://nvd.nist.gov/vuln/detail/CVE-2012-2459). Implicitly, trees are padded with empty leaves up to the smallest enclosing power of 2.
 
-For internal node with children `l` and `r`:
+For internal node with children `l` and `r`, its value `v` is:
 ```C++
-v = h(0x01, l, r) = h(0x01, l.v, r.v)
+v = h(l.v, r.v)
 ```
 
-## Sparse Binary Merkle Tree
+## Annotated Merkle Tree
+
+Merkle trees can be augmented as generic annotated Merkle trees, where additional fields can be contained in each node. One of the early annotated Merkle trees is the [Merkle Sum Tree](https://bitcointalk.org/index.php?topic=845978.0), which allows for compact fraud proofs to be made of fees collected in a block.
+
+Annotated Merkle trees have extra fields and methods to compute values for those fields, i.e. `f_1, ..., f_n, v` for `n` fields (note that if `n=0`, the annotated Merkle tree is a plain [binary Merkle tree](#binary-merkle-tree)). The value of field `f_i` is computed with the method `m_i_i(height, left_child_field, right_child_field)` for internal nodes and `m_i_l(message)` for leaf nodes.
 
+For leaf node of leaf message `m`, its value `v` and fields `f_1, ..., f_n` are:
+```C++
+f_1 = m_1_l(m)
+...
+f_n = m_n_l(m)
+v = h(serialize(m))
+```
+
+For internal node at height `height` with children `l` and `r`, its value `v` and fields `f_1, ..., f_n` are:
+```C++
+f_1 = m_1_i(height, l.f_1, r.f_1)
+...
+f_n = m_n_i(height, l.f_n, r.f_n)
+v = h(l.f_1, ..., l.f_n, l.v, r.f_1, ..., r.f_n, r.v)
+```
+
+If a compact Merkle root is needed, the root level (which consists of root fields and a root value) can be hashed once.
+
```

Comment:
Suggestion: Maybe add a sentence that (maybe how) annotating could also be used to differentiate between inner and leaf hashes?